File: leetcode_30days/07042020_counting_elements.rb
# frozen_string_literal: true

# @param {Integer[]} arr
# @return {Integer}
def count_elements(arr)
  dic = arr.each_with_object({}) { |num, o| o[num] = true }

  arr.inject(0) do |memo, num|
    next memo unless dic[num.succ]

    memo + 1
  end
end

File: leetcode_30days/07042020_counting_elements.js
var countElements = function(arr) {
  let numbers = arr.reduce((acc, num) => {
    acc[num] = true;
    return acc;
  }, {});

  return arr.reduce((count, num) => {
    if (numbers[num + 1]) {
        count += 1;
    }

    return count;
  }, 0);
};

var countElements = function(arr) {
  return arr.reduce((count, num) => {
    if (arr.indexOf(num + 1) > -1) {
        count += 1;
    }

    return count;
  }, 0);
};
